### PR TITLE
Add Span to various smir types

### PR DESCRIPTION
**Pull request by ouz-a**
Had to make few extra changes to few structs to attach spans to them.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_smir/src/rustc_smir/mod.rs [hunk 1]
```diff
@@ -225,39 +232,74 @@ impl<'tcx> Stable<'tcx> for mir::Statement<'tcx> {
...
-                fake_read_place.1.stable(tables),
-            ),
-            SetDiscriminant { place: plc, variant_index: idx } => {
-                stable_mir::mir::Statement::SetDiscriminant {
+            Assign(assign) => Statement {
+                kind: stable_mir::mir::StatementKind::Assign(
+                    assign.0.stable(tables),
+                    assign.1.stable(tables),
+                ),
+                span: self.source_info.span.stable(tables),
```
Comment (spastorino) [inline compiler/rustc_smir/src/rustc_smir/mod.rs:240]: I'd do `let span = self.source_info.span.stable(tables);` at the top and then just do `span,` here.

File: compiler/rustc_smir/src/rustc_smir/mod.rs [hunk 2]
```diff
@@ -808,52 +850,83 @@ impl<'tcx> Stable<'tcx> for mir::Terminator<'tcx> {
...
-                    .iter()
-                    .map(|(value, target)| stable_mir::mir::SwitchTarget {
-                        value,
-                        target: target.as_usize(),
-                    })
-                    .collect(),
-                otherwise: targets.otherwise().as_usize(),
+            Goto { target } => Terminator {
+                kind: TerminatorKind::Goto { target: target.as_usize() },
+                span: self.source_info.span.stable(tables),
```
Comment (spastorino) [inline compiler/rustc_smir/src/rustc_smir/mod.rs:857]: Same here, move the span assignment outside the match.

**Comment by spastorino** [comment]
This looks good to me.

**Comment by ouz-a** [comment]
> This looks good to me, modulo my tiny nit.

well code looks way cleaner now, thanks!

**Comment by spastorino** [comment]
Another thing that I was thinking that didn't realize yesterday, why don't we implement stable for kind?. Then I think you won't even need this span binding and code would likely be even cleaner.

**Comment by ouz-a** [comment]
> Another thing that I was thinking that didn't realize yesterday, why don't we implement stable for kind?. Then I think you won't even need this span binding and code would likely be even cleaner.

That could also work, and indeed will make the code cleaner but I think this will print out nicer Spans, and will make accessing spans easier, for parsing stuff in the future(if we do something like that).

**Comment by spastorino** [comment]
@ouz-a unsure if you've followed what I've said or if I'm missing something :). What I meant to say is equivalent (unless I'm missing something), you can just go for instance from `mir::StatementKind` to `stable_mir::mir::StatementKind` and those don't hold any span.

You would end with code like ...

```rust
Statement {
    kind: self.kind.stable(tables),
    span: self.source_info.span.stable(tables),
}
```

And adding the `StatementKind` thing.

**Comment by ouz-a** [comment]
> @ouz-a unsure if you've followed what I've said or if I'm missing something :). What I meant to say is equivalent (unless I'm missing something), you can just go for instance from `mir::StatementKind` to `stable_mir::mir::StatementKind` and those don't hold any span.
> 
> You would end with code like ...
> 
> ```rust
> Statement {
>     kind: self.kind.stable(tables),
>     span: self.source_info.span.stable(tables),
> }
> ```
> 
> And adding the `StatementKind` thing.

Oh I misread your message, I read that as implementing Span for Kinds, (it was late here when I read it 🤣 ) yeah your suggestion is right, I added it on last commit.

**Pull request merged**